python/pyarrow/tests/test_compute.py

```diff
@@ -1702,13 +1702,33 @@ def test_logical():
 
 
 def test_cast():
+    arr = pa.array([1, 2, 3, 4], type='int64')
+    options = pc.CastOptions(pa.int8())
+
+    with pytest.raises(ValueError):
+        pc.cast(arr, target_type=None)
+
+    with pytest.raises(ValueError):
+        pc.cast(arr, 'int32', safe=True, options=options)
+
+    with pytest.raises(ValueError):
+        pc.cast(arr, 'int32', safe=False, options=options)
+
+    assert pc.cast(arr, 'int8', options=options) == pa.array(
+        [1, 2, 3, 4], type='int8')
+
     arr = pa.array([2 ** 63 - 1], type='int64')
+    allow_overflow_options = pc.CastOptions(
+        pa.int32(), allow_int_overflow=True)
 
     with pytest.raises(pa.ArrowInvalid):
         pc.cast(arr, 'int32')
 
     assert pc.cast(arr, 'int32', safe=False) == pa.array([-1], type='int32')
 
+    assert pc.cast(arr, 'int32', options=allow_overflow_options) == pa.array(
+        [-1], type='int32')
```

Comment:
Now that I look at it…should we make the `type` an optional argument too? It's redundant with options (either you pass `type` and `safe`, or you pass `options`)

Reply:
Agreed. As it is now, you effectively pass the "to" type twice and error is returned if your value for `type` doesn't match that of `options.target_type`.